### PR TITLE
Restructure Map, Marker, and infoWindow objects

### PR DESCRIPTION
**Pull request by AliceVlasov**
Minimize direct interactions between classes without changing user experience.

tl;dr Overall functionality and user experience of the map did not change since code was mostly moved around and reorganized into new classes. The only new feature of the map is that users can also edit the location of a permanent marker, along with its title and body description, which was a side effect of the refactoring.

In order to minimize the direct interactions between permanent markers and temporary markers, the existing code was refactored such that the ChapMap instance handles all communication between markers and marker classes only handle their google.maps.Marker object and their relevant information window. 

In order to limit the number editable markers on the map, the edit feature of permanent markers was removed from the PermMarker/PermInfoWindow classs entirely and added to the TempMarker class so that only the map's unique temporary marker is able to receive information from the user which then has to be given to the ChapMap object in order to create or edit a permanent marker. To do this, a new TempInfoWindow class was defined which includes the same marker customization form for creating and editing permanent markers as before. Now, permanent markers don't have to keep track of whether or not it is being edited and then changing its information window content accordingly, and the temporary markers do not create new permanent markers directly. Since temporary markers are movable, this also allows clients to move permanent markers while they are being edited

Since the TempMarker and PermMarker classes had a lot of functions in common, the MarkerTemplate class was created as a super class for each of them to extend, so that only unique features of each marker needs to be kept in their class files. Similarly, the PermInfoWindow and TempInfoWindow classes now extent the InfoWindowTemplate class to minimize code duplication and to ensure controlled access and use.

Another change in structure was that, instead of the ChapMap having to handle the permanent information window, this was moved into the PermMarker class as a static property so now the ChapMap does not have to deal with any information windows directly.

### CODE REVIEW

File: src/main/webapp/ChapMap.js [hunk 1]
```diff
@@ -13,43 +13,136 @@
...
   }
 
   /**
    * @Private
-   * Removes the click listener allowing clients to add markers from the map
+   * Adds click listeners to the map customization buttons.
+   * A client can only add markers when the "adding markers" mode is on
+   */
+  addBtnListeners_(viewBtn, addMarkerBtn) {
+    addMarkerBtn.addEventListener('click', () => this.enableAddingMarkers_());
```
Comment (salehhamadeh) [inline src/main/webapp/ChapMap.js:69]: I like what you did here, with moving the methods to the class and calling them from an inline function.

File: src/main/webapp/MarkerTemplate.js [hunk 2]
```diff
@@ -12,19 +12,20 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+/**
+ * Defines the basic functions that all markers must have
+ * All classes that extend MarkerTemplate must implement:
+ *  - openMyInfoWindow()
```
Comment (salehhamadeh) [inline src/main/webapp/MarkerTemplate.js]: In JsDoc, this is done by defining empty methods and annotating them with `@abstract`.
Ex:
```
/**
 * Opens info window.
 *
 * @abstract
 */
openMyInfoWindow() {}
```

**Review by salehhamadeh — APPROVED** [review]

File: src/main/webapp/MarkerTemplate.js (hunk 2 above)
Comment (AliceVlasov) [reply src/main/webapp/MarkerTemplate.js]: Duly noted!

File: src/main/webapp/ChapMap.js (hunk 1 above)
Comment (AliceVlasov) [reply src/main/webapp/ChapMap.js:69]: Thank you! It's definitely a lot cleaner this way.

**Review by alexanderlopez — APPROVED** [review]

**Review by a-stepin — APPROVED** [review]
Lgtm!
